pkg/virt-handler/retry_manager.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package virthandler
+
+import (
+	"sync"
+	"time"
+)
+
+// FailRetryManager is the manger to handle asynchronous retry used by virt-handler.
+// When a failure event happens, virt-handler will try to fix it. The result of the
+// fix is reflected by whether we receive the failure signal again afterwards.
+// If we did, we want to schedule the retry in an exponential backoff manner.
+// This manager would remember the last try and return the wait time
+// for next try. If we don't receive the failure signal again within the
+// maxFailResponseTime, we consider the fix has worked and will reset the record.
+type FailRetryManager struct {
+	name string
+	// initialWait is the initial wait time for a retry. The wait time will be
+	// doubled afterwards for each retry until it reaches the maxWait time.
+	initialWait time.Duration
+	// maxWait is the max wait time between retries.
+	maxWait time.Duration
+	// maxFailResponseTime is the max time we expect to receive a failure signal
+	// after a retry. If no failure signal is received after maxFailResponseTime
+	// the retry is considered to have succeeded.
+	maxFailResponseTime time.Duration
+
+	stateLock   sync.Mutex
+	retryStates map[string]*retryState
+}
+
+type retryState struct {
+	firstFail     time.Time
+	lastRun       time.Time
+	nextRun       time.Time
+	waitInterval  time.Duration
+	lastRunFailed bool
+}
+
+// NewFailRetryManager creates a new FailRetryManager with the parameters explained above.
+func NewFailRetryManager(name string, initialWait, maxWait, maxFailResponseTime time.Duration) *FailRetryManager {
+	return &FailRetryManager{
+		name:                name,
+		maxFailResponseTime: maxFailResponseTime,
+		initialWait:         initialWait,
+		maxWait:             maxWait,
+		retryStates:         make(map[string]*retryState),
+	}
+}
+
+func (f *FailRetryManager) newState(now time.Time) *retryState {
+	return &retryState{
+		firstFail:     now,
+		lastRun:       now,
+		nextRun:       now.Add(f.initialWait),
+		waitInterval:  f.initialWait,
+		lastRunFailed: false,
+	}
+}
+
+// ShouldDelay returns whether the retry on this failure should be delayed or not,
+// and if true return the duration of the delay as well.
+func (f *FailRetryManager) ShouldDelay(key string, isFailure func() bool) (bool, time.Duration) {
```

Comment:
As a general note, this is an area that will be interesting for anyone who will be trying to troubleshoot these failures. I think these state changes should be logged.

Reply:
Thanks! Agreed, done!